internal/pkg/deploy/cloudformation/cloudformation.go

```diff
@@ -102,6 +126,54 @@ func (cf CloudFormation) streamResourceEvents(done <-chan struct{}, events chan
 	}
 }
 
+type renderStackChangesInput struct {
+	w                io.Writer
+	stackName        string
+	stackDescription string
+	createChangeSet  func() (string, error)
+	waitForStack     func(context.Context, string) error
+}
+
+func (cf CloudFormation) renderStackChanges(in renderStackChangesInput) error {
+	changeSetID, err := in.createChangeSet()
+	if err != nil {
+		return err
+	}
+	changeSet, err := cf.cfnClient.DescribeChangeSet(changeSetID)
+	if err != nil {
+		return err
+	}
+	body, err := cf.cfnClient.TemplateBody(in.stackName)
+	if err != nil {
+		return err
+	}
+	descriptions, err := cloudformation.ParseTemplateDescriptions(body)
+	if err != nil {
+		return fmt.Errorf("parse cloudformation template for resource descriptions: %w", err)
+	}
+
+	waitCtx, cancelWait := context.WithCancel(context.Background())
+	streamer := stream.NewStackStreamer(cf.cfnClient, in.stackName, changeSet.CreationTime)
+	renderer := progress.ListeningStackRenderer(streamer, in.stackName, in.stackDescription, resourcesToRender(changeSet.Changes, descriptions))
+
+	// Run the streamer, renderer, and waiter all concurrently until they all exit successfully or one of them errors.
+	// When the waiter exits, the waitCtx is canceled which results in the streamer and renderer to exit.
+	// When the streamer exits, the group ctx is canceled resulting in the waiter to exit and canceling the renderer.
+	// When the renderer exits, it's the same flow as the streamer.
+	g, ctx := errgroup.WithContext(waitCtx)
+	g.Go(func() error {
+		defer cancelWait()
+		return in.waitForStack(ctx, in.stackName)
+	})
+	g.Go(func() error {
+		return stream.Stream(waitCtx, streamer)
```

Comment:
Do we need to mock this one for testing?

Reply:
It's actually tested! See the test case: ["bubbles up streamer error and cancels waiter and renderer on streamer error"](https://github.com/aws/copilot-cli/pull/1793/files#diff-ff51f4486199743cf7adf97941b73ab33d0538bb6cd9e0db2fdc3aab2465d33eR102)